File: app/TumbleKit/Model/Photo.swift
import Foundation
import SwiftData

/// A single shot. Stored as SwiftData metadata; the pixels live on disk in the
/// shared container (see `PhotoStore`). A photo starts life *undeveloped* —
/// blank and face-down in the Drawer — until the shooter shakes it to life.
@Model
public final class Photo {
    /// Stable identity, also used to name the image files on disk.
    public var id: UUID = UUID()
    public var capturedAt: Date = Date()

    /// False until shake-to-develop finishes; controls the blank face-down state.
    public var isDeveloped: Bool = false
    /// 0…1 develop progress, persisted so a half-shaken print resumes where it was.
    public var developProgress: Double = 0

    /// File names within the shared images directory.
    public var rawImageName: String?
    public var developedImageName: String?

    /// Drawer placement, generated once at capture so the scatter is stable
    /// across launches (never a grid). Percent offsets + rotation in degrees.
    public var scatterX: Double = 0
    public var scatterY: Double = 0
    public var rotation: Double = 0

    public var caption: String?
    /// Where the shot came from: the app or the lock-screen extension.
    public var source: String = PhotoSource.app.rawValue

    public init(
        id: UUID = UUID(),
        capturedAt: Date = Date(),
        source: PhotoSource = .app
    ) {
        self.id = id
        self.capturedAt = capturedAt
        self.source = source.rawValue
        let scatter = Photo.randomScatter()
        self.scatterX = scatter.x
        self.scatterY = scatter.y
        self.rotation = scatter.rotation
    }

    /// A loose, hand-tossed placement in the drawer area (percent offsets,
    /// gentle rotation) — echoes the site's `DrawerMockup` scatter.
    static func randomScatter() -> (x: Double, y: Double, rotation: Double) {
        (
            x: Double.random(in: 2...52),
            y: Double.random(in: 2...70),
            rotation: Double.random(in: -12...12)
        )
    }
}

public enum PhotoSource: String, Sendable {
    case app
    case lockscreen
}

public extension Photo {
    /// How aged the print looks, 0 (fresh) → 1 (fully warmed/faded), mapped
    /// over `agingSpan`. Purely a function of elapsed time — no stored state,
    /// so prints visibly warm as the days pass.
    static let agingSpan: TimeInterval = 60 * 60 * 24 * 30 // ~30 days

    func ageFraction(now: Date = Date()) -> Double {
        let elapsed = now.timeIntervalSince(capturedAt)
        return min(1, max(0, elapsed / Photo.agingSpan))
    }
}

File: android/app/src/main/java/com/tumble/model/Photo.kt
package com.tumble.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.time.Duration
import java.time.Instant
import java.util.UUID
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

enum class PhotoSource(val value: String) {
    APP("app"),
    LOCKSCREEN("lockscreen"),
}

/**
 * A single shot. Stored as Room metadata; the pixels live on disk (see
 * `PhotoStore`). A photo starts life *undeveloped* — blank and face-down in the
 * Drawer — until the shooter shakes it to life.
 *
 * Ported from `app/TumbleKit/Model/Photo.swift`.
 */
@Entity(tableName = "photos")
data class Photo(
    /** Stable identity, also used to name the image files on disk. */
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val capturedAt: Instant = Instant.now(),

    /** False until shake-to-develop finishes; controls the blank face-down state. */
    val isDeveloped: Boolean = false,
    /** 0…1 develop progress, persisted so a half-shaken print resumes where it was. */
    val developProgress: Double = 0.0,

    /** File names within the app images directory. */
    val rawImageName: String? = null,
    val developedImageName: String? = null,

    /**
     * Drawer placement, generated once at capture so the scatter is stable
     * across launches (never a grid). Percent offsets + rotation in degrees.
     */
    val scatterX: Double = 0.0,
    val scatterY: Double = 0.0,
    val rotation: Double = 0.0,

    val caption: String? = null,
    /** Where the shot came from: the app or the lock-screen surface. */
    val source: String = PhotoSource.APP.value,
) {
    companion object {
        /** How aged the print looks maps over ~30 days. */
        val agingSpan: Duration = Duration.ofDays(30)

        /**
         * A loose, hand-tossed placement in the drawer area (percent offsets,
         * gentle rotation) — echoes the site's `DrawerMockup` scatter.
         */
        fun randomScatter(random: Random = Random.Default): Triple<Double, Double, Double> =
            Triple(
                random.nextDouble(2.0, 52.0),
                random.nextDouble(2.0, 70.0),
                random.nextDouble(-12.0, 12.0),
            )

        /** Mirror of the Swift `init` — draws a stable scatter at capture time. */
        fun create(
            id: String = UUID.randomUUID().toString(),
            capturedAt: Instant = Instant.now(),
            source: PhotoSource = PhotoSource.APP,
            random: Random = Random.Default,
        ): Photo {
            val (x, y, rotation) = randomScatter(random)
            return Photo(
                id = id,
                capturedAt = capturedAt,
                scatterX = x,
                scatterY = y,
                rotation = rotation,
                source = source.value,
            )
        }
    }
}

/**
 * How aged the print looks, 0 (fresh) → 1 (fully warmed/faded), mapped over
 * `agingSpan`. Purely a function of elapsed time — no stored state, so prints
 * visibly warm as the days pass.
 */
fun Photo.ageFraction(now: Instant = Instant.now()): Double {
    val elapsedMs = Duration.between(capturedAt, now).toMillis().toDouble()
    val spanMs = Photo.agingSpan.toMillis().toDouble()
    return min(1.0, max(0.0, elapsedMs / spanMs))
}
